segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/metadata/BaseMetadataStore.java

```diff
@@ -527,6 +538,138 @@ protected void setVersion(long version) {
         this.version.set(version);
     }
 
+    /**
+     * Helper class to encapsulates lock over group of keys.
+     */
+    private static class LockWrapper {
+        /**
+         * Global map to track active locks.
+         */
+        private static final HashMap<String, LockData> LOCKS = new HashMap<>();
+
+        /**
+         * Keys to lock.
+         */
+        private final String[] keysToLock;
+
+        /**
+         * Keeps track of acquired locks.
+         */
+        private final Lock[] locks;
+
+        /**
+         * Keeps track of number of locks are acquired.
+         */
+        private int aquiredLockCount = 0;
+
+        /**
+         * Whether to acquire read locks or write locks.
+         */
+        private boolean isReadOnly;
+
+        /**
+         * Constructs a {@link LockWrapper} instance.
+         * @param isReadOnly Whether to acquire read only locks or writable locks.
+         * @param keysToLock Keys to lock.
+         */
+        LockWrapper(Boolean isReadOnly, String... keysToLock) {
+            Preconditions.checkState(keysToLock.length > 0, "At least one key must be locked.");
+            this.keysToLock = keysToLock;
+            this.locks = new Lock[keysToLock.length];
+            this.isReadOnly = isReadOnly;
+        }
+
+        /**
+         * Creates a LockWrapper for reading.
+         * @param keysToLock Keys to lock.
+         * @return LockWrapper instance.
+         */
+        static LockWrapper readLock(String... keysToLock) {
+            return new LockWrapper(true, keysToLock);
+        }
+
+        /**
+         * Creates a LockWrapper for writing.
+         * @param keysToLock Keys to lock.
+         * @return LockWrapper instance.
+         */
+        static LockWrapper writeLock(String... keysToLock) {
+            return new LockWrapper(false, keysToLock);
+        }
+
+        /**
+         * Acquires all required locks.
+         */
+        void lock() {
+            for (val key: keysToLock) {
```

Comment:
Is there any guarantee on the ordering of the elements of this array?

If we do not set a specific order we can fall into deadlocks

Reply:
In reality almost all operations use only one key. 
And the only operation that uses two keys is Storage::concat which is the last operation on source segment.
So in practice unlikely.
But yes will enforce some order.

Reply:
No acquiring locks in predefined order (segment name sort order).  Releasing in reverse order.